ENH: stats.multivariate_t: add cdf method

mdhaber opened the pull request:
#### Reference issue
[gh-16241](https://github.com/scipy/scipy/pull/16241/commits/20d88ae7c47d5394bd83fd897dc919d437c0d93e#r881819926)

#### What does this implement/fix?
This adds a CDF method to `multivariate_t`.

#### Additional information
Rather than wrapping Genz's [Fortran code](https://www.math.wsu.edu/faculty/genz/software/software.html), I translated his (permissively-licensed) [Matlab code](https://www.math.wsu.edu/faculty/genz/software/matlab/qsimvtv.m) so that it would be easier to use a NumPy's PRNG (and avoid the equivalent of gh-16142). 

That said, the number of random variates needed is always `10*dim`. My plan was to compile this with Pythran in a follow-up PR, and to do that I would have to pass in those random variates because I don't think we can pass in a `Generator` or `RandomState` into a Pythran function (correct me if I'm wrong). In that case, I suppose we could apply the same strategy with the Fortran code, slightly modified to accept all the random variates as an additional input. The advantage is that we wouldn't really need to review the Fortran source; the disadvantage is that we'd need to wrap it.

I still need to look for rough edges and polish them. I suspect that the output shapes are not perfect, I don't know what happens when `x` is empty, etc.  *Update: it handles edge cases like `multivariate_t.pdf` does.*

I went ahead and wrote my own translation because comparing existing translations may be a useful part of review.

rkern commented:
Ah, I should have made my translation public earlier. I've been sitting on it for a while.

https://gist.github.com/rkern/8518cb3e3f57fa81d21dbd9eabcfaa87

Yeah, lifting the drawing from `Generator` out of the loop is pretty easy. Then the inner loop would be Pythranizable. The main blocker then is adding the necessary special functions to Pythran. I'm not sure you can do that except by releasing a new version of Pythran with those implemented.

The CBC lattice in [Genz's later version](https://www.math.wsu.edu/faculty/genz/software/matlab/qsilatmvtv.m) is quite a bit superior to the Richtmyer lattice. @tupui I don't know if you want to consider adding the CBC lattice as a generic QMC engine. It's finicky (more than Sobol' even) in that you have to draw exactly the prime number of samples that it was configured to produce. The core algorithm is here to use, if you want.

I have an implementation of the [generic multivariate normal _integrand_](https://gist.github.com/rkern/8518cb3e3f57fa81d21dbd9eabcfaa87#file-qmvnt-py-L241) that you can integrate yourself with any `QMCEngine` to compare techniques. In my findings, the CBC lattice here nearly always edges out Sobol' in terms of accuracy and runtime, but they are not far.

If you want a test problem, integrating a multivariate normal with the covariance matrix with the diagonals all 1 and the off-diagonals all 0.5 over the negative orthant has an exact true answer (`1 / (ndim + 1)`). And it's a nontrivial covariance matrix that appears in real life problems (Dunnett's test with equal group sizes).

mdhaber commented:
Thanks! Yeah, I knew you already translated it, but I figured one could be tested against the other. Or maybe we'll end up going for Fortran anyway. 
I'll add that test problem, thanks!

@serge-sans-paille any chance we can get `special.gammaincinv` and `ndtri` in Pythran? (And ideally `ndtr`, but it looks like we can do that with `erf` if need be.)

rkern commented:
I believe that the CBC QMC scheme used in [the version that I translated](https://www.math.wsu.edu/faculty/genz/software/matlab/qsilatmvtv.m) is superior in performance to the scheme in [the Fortan code](https://www.math.wsu.edu/faculty/genz/software/fort90/mvdist.f90). I think Pythranization is the way to go.

We should probably wrap the [Fortran code for the bivariate and trivariate cases](https://www.math.wsu.edu/faculty/genz/software/fort77/tvpack.f) which can be done more accurately.

mdhaber commented:
> I think Pythranization is the way to go.

OK. I like that if we can get the special functions. If not, there's always Cython. 

> We should probably wrap the Fortran code for the bivariate and trivariate cases which can be done more accurately.

Good idea. But let's leave that as a followup PR. This will get big enough as it is.

---

I had planned to wait until @tupui get's back next month before doing too much more here, but I suppose I should ask - would you like to review this, or would you like to submit your code as a PR and I will review? 

In either case, I'd like to suggest that given the same input (including random numbers), the translation should produce the same output up to machine precision. As long as we hit all the code paths with tests, this would make it very easy to check for accuracy against the Matlab originals. Is that still true in your translation?

rkern commented:
Not quite. The CBC lattice construction involves doing some convolutions via FFTs and `argmin()` on the results. Sometimes there are near ties (i.e. would be true ties in infinite-precision arithmetic but aren't in `float64`). Due to the way MATLAB sometimes rounds small numbers down to 0, sometimes it breaks the near tie in a different way than we do. They are equally valid lattices but will cause different results. I was checking my results against the MATLAB code as evaluated by Octave and discovered that.

serge-sans-paille commented:
> @serge-sans-paille any chance we can get `special.gammaincinv` and `ndtri` in Pythran? (And ideally `ndtr`, but it looks like we can do that with `erf` if need be.)

I can give it a try. Can you open an issue in Pythran bug tracker to keep note of it?

serge-sans-paille commented:
@mdhaber I didn't plan to do a pythran release soon, but once you have that one ready, please tell me and I'll do a minor version release.

mdhaber commented:
Thanks @serge-sans-paille!

mdhaber commented:
Welcome back @tupui! What would you prefer to do here?

1. Continue with Genz's `qsimvtv` (what we have here).
2. Have me translate Genz's `qsilatmvtv` (improved) in the same style (i.e. roughly line-by-line translation).
3. Start with 1, and do 2 in a followup.
4. Replace this with Robert's translation.

tupui commented:
Thanks 😃

I still need to have a look and make a decision. Was busy with other reviews so far. I will try to have an answer tmr.

tupui commented:
Sorry for the delay.

> Welcome back @tupui! What would you prefer to do here?
> 
> 1. Continue with Genz's `qsimvtv` (what we have here).
> 2. Have me translate Genz's `qsilatmvtv` (improved) in the same style (i.e. roughly line-by-line translation).
> 3. Start with 1, and do 2 in a followup.
> 4. Replace this with Robert's translation.

What about we use @rkern's version with CBC and use your current version as a validation test?

mdhaber commented:
OK @tupui I updated per your recommendation. Those Cirrus failure seems unrelated. Other failures seem to be due to an fftpack import; I'll see if I can switch that out.

@rkern how do you want to be credited? Before merging, we could do an interactive rebase in which you would author the first commit (maybe just adding the file `_qmvnt.py`) and we'd squash all the rest of my commits on top. Alternatively, would you be ok being listed as a coauthor when we squash-merge (easier)?

mdhaber commented on scipy/stats/_qmvnt.py:
```diff
@@ -0,0 +1,506 @@
+
+import numpy as np
+
+from scipy.fftpack import fft, ifft
```
```suggestion
from scipy.fft import fft, ifft
```
This should fix the FFT issue. I tested it locally and it works.

rkern commented:
Whatever's easiest.

tupui commented on scipy/stats/_qmvnt.py:
```diff
@@ -0,0 +1,506 @@
+    low = np.asarray(low, dtype=np.float64)
+    high = np.asarray(high, dtype=np.float64)
+    cho, lo, hi = _permuted_cholesky(covar, low / sn, high / sn)
+    n = cho.shape[0]
```
`n` is supposed to be the dimension no? If so we should rename it.

tupui commented on scipy/stats/_qmvnt.py:
```diff
@@ -0,0 +1,506 @@
+        pv = np.ones(n_qmc_samples)
+        s = np.zeros((n, n_qmc_samples))
+        for i in range(n):
+            # Pseudorandomly-shifted lattice coordinate.
+            z = q[i] * i_samples + rng.random()
+            # Fast remainder(z, 1.0)
+            z -= z.astype(int)
+            # Tent periodization transform.
+            x = abs(2 * z - 1)
```
I am really confused by this thing. Seems like the intent is to have a sample which is either quite regular with CBC or Richtmyer. Why not just using a QMC sample? This is an integration problem right?

rkern replied:
As I said earlier, CBC nearly always seems to give more accurate results faster than Sobol'.

tupui replied:
Even as the dimensionality increases? I don't understand why, the integrand should be smooth here no? Which would be the perfect scenario for (R)QMC.

Did you use `2**n` points for Sobol'? Did you try Halton or LHS?

rkern replied:
You can play with the notebook in [the gist I linked above](https://gist.github.com/rkern/8518cb3e3f57fa81d21dbd9eabcfaa87) to investigate on your own.

Yes, even as the dimensionality increases; Sobol' approaches CBC, but never really beats it (except for the occasional random fluctuation).

Yes, I used perfect `2**n` points for Sobol'.

Yes, I tried Halton and LHS. Both are significantly less accurate compared to Sobol'.

By speed, I am referring to the wall-clock time for computing the CDF (so generating the QMC points and evaluating the integral). Since the accuracy of the result is never better than CBC for the same number of points (modulo the rounding to the nearest prime when necessary), it always takes at least as many evaluations, so also "slower" in the convergence sense.

rkern replied:
The integrand also has at least one singularity on the boundary, which the CBC lattice always avoids, but Sobol' sometimes doesn't, depending on the scrambling.

rkern replied:
Also, not for nothing, the methods implemented here _are_ RQMC, as [well-explained by our own documentation introducing QMC](http://scipy.github.io/devdocs/reference/stats.qmc.html#introduction-to-quasi-monte-carlo).

tupui replied:
> Yes, even as the dimensionality increases; Sobol' approaches CBC, but never really beats it (except for the occasional random fluctuation).
> ...
> The integrand also has at least one singularity on the boundary, which the CBC lattice always avoids, but Sobol' sometimes doesn't, depending on the scrambling.

If there are known points to avoid, we can and should work around them. Otherwise it's not a fair comparison.

I will try to run your notebook.

> Also, not for nothing, the methods implemented here are RQMC, as well-explained by our own documentation introducing QMC.

I know, I wrote that with Art 😉 Sobol' is just supposed to converge faster on well behaved integrand.

Not sure what you mean by, not for nothing. As it is coded at the moment it's not really following our framework.

rkern replied:
> If there are known points to avoid, we can and should work around them. Otherwise it's not a fair comparison.

I guess so. I don't see how that's going to improve the performance of Sobol' in this comparison, though. Whatever you do to avoid the `inf` will likely be at a performance cost for no real improvement in the accuracy over CBC.

I guess you could try to derive a different transformation of the integrand that doesn't have the same boundary issue, but the goal here isn't to make "fair" comparisons between CBC lattices and Sobol' generators overall. It's to implement multivariate-t CDFs.

> Not sure what you mean by, not for nothing. As it is coded at the moment it's not really following our framework.

I wanted to be clear to everyone else about what you were claiming. The way that you wrote "Why not just using a QMC sample?" and "Which would be the perfect scenario for (R)QMC." it appeared that you were claiming that what was implemented was not RQMC at all because it was a lattice rule rather than one of the other methods. I wanted it to be clear to everyone that yes, these are RQMC algorithms. Ones that are well-validated for this particular problem by [the guy who literally wrote the book on the subject](https://link.springer.com/book/10.1007/978-3-642-01689-9). If you want to ask about why I'm not using "existing `QMCEngine` implementations" or "`scipy.stats.qmc` framework", that would be a better phrasing. What's currently implemented is not the whole landscape of reasonable options in the domain.

In my experiments, the existing `QMCEngine`s can only come close to the CBC implementation for this integrand, but do not exceed it as an algorithmic matter (i.e. numerical accuracy and sample efficiency). When we get to the Pythranization or other implementation optimizations, the lattice implementation will likely only get better in comparison due to implementation efficiencies because point generation is inlined in the loop very efficiently. Even if we added a `CBCLattice` `QMCEngine`, I probably wouldn't want to use `qmc_quad()` with it in any case (though I would make sure that I could reuse its lattice generator-finding implementation).

rkern replied:
One can reintroduce some guarding logic from the original MATLAB code in order to restore good behavior on the boundary. It doesn't change anything about the performance of `qmc_quad()` with any of the `QMCEngine`s, though, just guards against the internal `NaN`s from leaking.

Replace

```python
            c = phi((lo[i] - s) / ct)
            d = phi((hi[i] - s) / ct)
```

with 

```python
            c = np.ones(n_qmc_samples)
            d = np.ones(n_qmc_samples)
            lois = lo[i] - s
            hiis = hi[i] - s
            c[lois < -9*ct] = 0.0
            d[hiis < -9*ct] = 0.0
            lo_mask = abs(lois) < 9*ct
            hi_mask = abs(hiis) < 9*ct
            c[lo_mask] = phi(lois[lo_mask] / ct)
            d[hi_mask] = phi(hiis[hi_mask] / ct)
```

in the various loops.

mdhaber replied:
What is `ct`?

When the covariance is singular, I've found that it's not uncommon for `d` below to be zero.  This leads to `y` being `-inf`. Is this the singularity mentioned above?

Without thinking about what is really going on, I tried fixing it by calculating `dc` with `log_ndtr` and `logsumexp`, but then that can underflow. I think the whole calculation can be moved to logspace with the added help of `ndtri_exp`, but it sounds like there may be another option?

rkern replied:
It has nothing to do with the PDF being degenerate. The particular transformation of the PDF to the unit hypercube just has a cusp at the unit boundaries in some circumstances (particularly when the regular-space bounds are semi-infinite) for general covariance matrices. The naive evaluation yields `lo[i] = -inf` and `s = -inf` creating a `NaN` when they are subtracted. The MATLAB code's extra masking stops the `NaN` from leaking out and replacing it with a 0, but the actual value that comes out is not quite correct for the cusp (which does seem to have a finite, nonzero limiting value).

tupui replied:
> the goal here isn't to make "fair" comparisons between CBC lattices and Sobol' generators overall. It's to implement multivariate-t CDFs.

A nuance here, the goal is to implement multivariate-t CDF in a maintainable way that maintainers who will need to fix bug can understand and fix.

Also note that the usual "social/progamming" contract we tend to have-at the moment at least-is that when claims are made, it's not up to the reviewer to do the work of proving them. We are going in much greater length on some topics to verify and prove things. And sometimes it even involves authors of papers. Art even wrote a paper after the QMC PR. And we could write one as well with what was done for Boschloo/Barnard. I am not saying we should write a paper for every PR OC and there is a limit to what is reasonable to be asked.

Here you are proposing 2 other QMC methods instead of the one we already have. I think it's legitimate to ask why we should add other methods just for this purpose and not follow our existing infrastructure.

If in this case you are certain it's faster and converge faster. Then ok we have our answer and can accept the additional maintenance cost knowingly.

> When we get to the Pythranization or other implementation optimizations, the lattice implementation will likely only get better in comparison due to implementation efficiencies because point generation is inlined in the loop very efficiently.

I am not sure if pure speed is very relevant here, but ok.

Just a note that our QMC engines are not that slow as being implemented in Cython (https://gist.github.com/tupui/54a57f474a069f685ba85c9036f6b121)

> Even if we added a CBCLattice QMCEngine, I probably wouldn't want to use qmc_quad() with it in any case (though I would make sure that I could reuse its lattice generator-finding implementation).

As said above, fine by me if you maintain this code.

I am personally more familiar with the use of the engines and `qmc_quad`. I don't get the interlacing between samples generation, integration and the rest, but that's fine.

tupui commented on scipy/stats/_qmvnt.py:
```diff
@@ -0,0 +1,506 @@
+    return prob, est_error, n_samples
+
+
+def _bounds(m, t, alternative):
```
Provision?

rkern replied:
It's for the Dunnett test. Should be removed from here because I don't think this file will be where that gets implemented.

tupui reviewed (commented):
Sorry for the delay.

I did a first pass. I am not sure about the methods we need in the end in `_qmvnt.py`. There are lot of unused code so before I dive into some function it would help to clarify what is to stay.

rkern commented on scipy/stats/_qmvnt.py:
```diff
@@ -0,0 +1,506 @@
+    return prob, est_error, n_samples
+
+
+def _qmvn(m, covar, low, high, rng, lattice='cbc', n_batches=10):
+    """Multivariate normal integration over box bounds.
```
This is likely to get used later to fix #16142

tupui replied:
ok but if not needed for now, we should remove it to ease the review.

rkern replied:
It's also a good way to evaluate different implementation choices, like the CBC versus Sobol'. We have some nontrivial examples multivariate normal integrations with known exact answers. There aren't as many for multivariate-t.

rkern commented on scipy/stats/_qmvnt.py:
```diff
@@ -0,0 +1,506 @@
+    return prob, est_error, n_samples
+
+
+def _mvn_qmc_integrand(covar, low, high, use_tent=False):
```
More of a leftover from my standalone version which was used to explore the properties and compare alternatives. I used this to show that CBC worked better for this integrand than Sobol', for instance.

rkern commented on scipy/stats/tests/data/_mvt.py:
```diff
@@ -0,0 +1,168 @@
+
+
+def _primes(n):
+    return primes_from_2_to(math.ceil(n))
```
No, arguably the `math.ceil()` should be in-lined at the one call-site. Whether a floating point number should be rounded up or down is very application specific.

rkern commented on scipy/stats/_qmvnt.py:
```diff
@@ -0,0 +1,506 @@
+    return lattice(n_dim, n_qmc_samples)
+
+
+def _qauto(func, covar, low, high, rng, error=1e-3, limit=10_000, **kwds):
```
It would probably be good to implement `multivariate_t.cdf()` to use this so that users can specify their desired accuracy rather than try to guess at the number of points necessary.

mdhaber commented on scipy/stats/tests/data/_mvt.py:
```diff
@@ -0,0 +1,168 @@
+    # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+    # Initialization
+    sn = max(1, math.sqrt(nu)); ch, az, bz = _chlrps(sigma, a/sn, b/sn)
```
I used `z` instead of "s" because `as` is reserved.

rkern commented on scipy/stats/_qmvnt.py:
```diff
@@ -0,0 +1,506 @@
+    return q, n_qmc_samples
+
+
+def _factorize_int(n):
```
```suggestion
def _factorize_int(n):
    """Return a sorted list of the unique prime factors of a positive integer.
    """
```

rkern commented on scipy/stats/_qmvnt.py:
```diff
@@ -0,0 +1,506 @@
+    return sorted(factors)
+
+
+def _primitive_root(p):
```
```suggestion
def _primitive_root(p):
    """Compute a primitive root of the prime number `p`.
    
    Used in the CBC lattice construction.
    
    References
    ----------------
    .. [1] https://en.wikipedia.org/wiki/Primitive_root_modulo_n
    """
```

rkern commented on scipy/stats/_qmvnt.py:
```diff
@@ -0,0 +1,506 @@
+    return r
+
+
+def _cbc_lattice(n_dim, n_qmc_samples):
```
```suggestion
def _cbc_lattice(n_dim, n_qmc_samples):
    """Compute a QMC lattice generator using a Fast CBC construction.

    Parameters
    ----------
    n_dim : int > 0
        The number of dimensions for the lattice.
    n_qmc_samples : int > 0
        The desired number of QMC samples. This will be rounded down to the nearest
        prime to enable the CBC construction.

    Returns
    -------
    q : float array : shape=(n_dim,)
        The lattice generator vector. All values are in the open interval `(0, 1)`.
    actual_n_qmc_samples : int
        The prime number of QMC samples that must be used with this lattice, no more,
        no less.

    References
    ----------
    .. [1] Nuyens, D. and Cools, R. "Fast Component-by-Component Construction, a
           Reprise for Different Kernels", In H. Niederreiter and D. Talay,
           editors, Monte-Carlo and Quasi-Monte Carlo Methods 2004,
           Springer-Verlag, 2006, 371-385.
    """
```

mdhaber commented on scipy/stats/_qmvnt.py:
```diff
@@ -0,0 +1,491 @@
+                    r = np.ones_like(x)
+            else:
+                y = phinv(c + x * dc)  # noqa: F821
+                s[i:, :] += cho[i:, i - 1][:, np.newaxis] * y
```
Does this look safe to you? The problem I mentioned above was about `y` being infinite, and this calculation gives an "invalid" warning. (But it does not make the output NaN.)

```suggestion
                with np.errstate(invalid='ignore'):
                    s[i:, :] += cho[i:, i - 1][:, np.newaxis] * y
```

mdhaber commented on scipy/stats/_qmvnt.py:
```diff
@@ -0,0 +1,491 @@
+                s[i:, :] += cho[i:, i - 1][:, np.newaxis] * y
+            si = s[i, :]
+
+            c = phi(lo[i] * r - si)
+            d = phi(hi[i] * r - si)
```
Is this what you had in mind?
```suggestion
            ct = cho[i, i]
            c = np.ones(n_qmc_samples)
            d = np.ones(n_qmc_samples)
            with np.errstate(invalid='ignore'):
                lois = lo[i] * r - si
                hiis = hi[i] * r - si
            c[lois < -9*ct] = 0.0
            d[hiis < -9*ct] = 0.0
            lo_mask = abs(lois) < 9*ct
            hi_mask = abs(hiis) < 9*ct
            c[lo_mask] = phi(lois[lo_mask] / ct)
            d[hi_mask] = phi(hiis[hi_mask] / ct)
```

rkern replied:
Sorry, I wrote that snippet for the multivariate normal case, since that's the one that I had the integrand function set up for testing with `qmc_quad()`, which was the point under discussion. You don't deal with `ct` in the multivariate-t code.

mdhaber reviewed (commented):
I'm getting warnings when the covariance matrix is singular.

> The MATLAB code's extra masking stops the NaN from leaking out and replacing it with a 0, but the actual value that comes out is not quite correct for the cusp (which does seem to have a finite, nonzero limiting value).

So should we warn the user about a loss of accuracy then?

For singular covariance, I'm also getting a lot of substantial discrepancies between this and `_qsimvtv` without much improvement when I increase the number of points in `_qsimvtv`. Would you expect `_qsimvtv` to be very inaccurate in this case?

rkern commented:
> So should we warn the user about a loss of accuracy then?

What I was talking about is not a case that arises with the CBC lattice implemented here. It would arise if we did use `qmc_quad()` with Sobol'. There are several other reasons why I don't think that would be an improvement, regardless of the boundary issue. In any case, I don't think there's any significant loss of accuracy in any case in which the mask is triggered.

rkern commented:
> Would you expect `_qsimvtv` to be very inaccurate in this case?

I don't expect any implementation to be particularly accurate in the singular case.

rkern commented:
I've updated [my gist with a notebook with some more test problems](https://gist.github.com/rkern/8518cb3e3f57fa81d21dbd9eabcfaa87#file-mvtestproblems-ipynb). It turns out that the analytically-known case that I used in my benchmark is a special case of a slightly wider family of a structured covariance matrix where the multivariate normal integral can be reduced down to a univariate integral that is straightforward to numerically integrate with `quad()`.

A related special case family creates a singular matrix, and it too can be integrated numerically.

And finally, it does turn out that one similar special case family can turn the multivariate t into a double-integral that `dblquad()` can handle fine. I think it is likely that the strategy used in the singular MVN case can be ported over to the singular MVT case if we really care about it, but I'm definitely not going through the math to derive it.

tupui reviewed (commented):
> I've updated [my gist with a notebook with some more test problems](https://gist.github.com/rkern/8518cb3e3f57fa81d21dbd9eabcfaa87#file-mvtestproblems-ipynb). It turns out that the analytically-known case that I used in my benchmark is a special case of a slightly wider family of a structured covariance matrix where the multivariate normal integral can be reduced down to a univariate integral that is straightforward to numerically integrate with quad().

I am not sure I understand what conclusion you want to draw.

I am not going to block this. It's in relative ok shape and main points have been discussed. Let's finish up last ones and we can merge this soon.

rkern commented:
> I am not sure I understand what conclusion you want to draw.

That notebook has no particular conclusion. I am offering it as an additional source of test problems to be used in our test suite rather than testing by comparing two ports of the only slightly-different versions of basically the same MATLAB code.

It can also be used to do more thorough comparisons on the relative accuracy between the CBC lattice and Sobol' on a wider variety of problems. I ran those benchmarks last night. My findings from the more restricted problem hold up. If there is interest, I can clean up the visualizations of those results.

> Here you are proposing 2 other QMC methods instead of the one we already have. I think it's legitimate to ask why we should add other methods just for this purpose and not follow our existing infrastructure.

Just one, we can drop the Richtmyer lattice; it's always worse.

On the whole, I have not questioned the legitimacy of your questions. I have simply answered them. I anticipated the question about using `qmc_quad()` with Sobol' (I was curious myself, and hoping that it would be better!), which is why I did the comparison ahead of this PR and published the testing code before you asked your questions.

The only thing I did question was the relevancy of one complaint about the fairness of how the comparison was done with respect to the boundary issue, most particularly because there was no mechanism by which restoring "fairness" on that particular issue could work in Sobol''s favor in terms of accuracy or time. It was a secondary issue that could not affect the primary conclusion.

In the event, it didn't matter. All combinations (doing the fastest thing for both, doing the safest thing for both even if it was unnecessary, doing the optimal thing for both) don't change the relative orderings of accuracy and time.

> If in this case you are certain it's faster and converge faster. Then ok we have our answer and can accept the additional maintenance cost knowingly.

I am very confident of this, particularly with the new benchmark problems. Yes, I will maintain this code. The inline generation of the QMC points is, by a long stretch, the easiest part of this code to understand and maintain.

> Just a note that our QMC engines are not that slow as being implemented in Cython

For sure! They are not slow, by any means. Their API, though it is a very good general purpose API, is just less well-adapted to this particular problem. The integrand is arranged such that it can deal with each dimension one by one, vectorizing across all of the samples. I don't need to generate whole blocks of QMC samples, just one column at a time. Lattice rules work really well for that. But even if we did have lattice `QMCEngine`s, their APIs and that of `qmc_quad()` doesn't let us make use of that.

tupui commented:
> The integrand is arranged such that it can deal with each dimension one by one, vectorizing across all of the samples. I don't need to generate whole blocks of QMC samples, just one column at a time. Lattice rules work really well for that. But even if we did have lattice QMCEngines, their APIs and that of qmc_quad() doesn't let us make use of that.

This is something I have in the back of my mind, add a dimension to a sample/engine. A lot of QMC methods, Sobol' to start with, allow to add a dimension (and are built like that like here.) I am mostly waiting for more people to ask this and really put some thoughts on that. I also have a use case for some people doing DoE but that might be too niche (and also arguable) for now.

mdhaber commented:
I fixed the PEP8 issues in the docstrings and removed or commented about the unused code. I also added tests against generic integrators, Matlab, and the analytical result for the singular case mentioned in https://github.com/scipy/scipy/pull/17410#issuecomment-1312628547.

rkern commented on scipy/stats/_qmvnt.py:
```diff
@@ -222,7 +216,7 @@ def _qmvn(m, covar, low, high, rng, lattice='cbc', n_batches=10):
     rng : Generator, optional
         default_rng(), yada, yada
-    lattice : 'cbc' or 'richtmyer' or callable
+    lattice : 'cbc' or callable
```
This parameter is no longer used. It can be removed from the signature and docstring.

tupui commented on scipy/stats/tests/test_multivariate.py:
```diff
@@ -2348,6 +2364,174 @@ def test_rvs(self, size, dim, df):
+            ref = _qsimvtv(20000, df, cov, a - mean, b - mean, rng)[0]
+        assert_allclose(res, ref, atol=1e-4, rtol=1e-3)
+
+    def test_cdf_against_generic_integrators(self):
```
+1, I had a very similar code locally when doing some tests. I also had a simple 1D case with an empirical CDF but this should be more robust and cover this.

tupui reviewed (approved):
LGTM and CI is happy. Let's get this in. Thanks all for working on that and the discussions. (Unrelated failures on Cirrus as too usual.)

rkern commented:
> the analytical result for the singular case mentioned in [#17410 (comment)](https://github.com/scipy/scipy/pull/17410#issuecomment-1312628547).

Just for the record, this case is not singular.

Also, I presented it as an analytic result for the multivariate normal. Fortunately, it also seems to be true for the multivariate t, regardless of `df`, at least numerically. I don't have a derivation for it, though, like I do for the multivariate normal case.

rkern commented:
Ah, I do have a derivation. For the negative orthant bounds, all of the `w` terms that interact with `df` disappear. The `w` terms can all be moved outside of the `u` integral and evaluate to 1. Then the remaining integrand reduces to the same gaussian integral from the corresponding multivariate normal case.

mdhaber commented:
@rkern there is an effort to remove very old Fortran 77 code (gh-18566) and one of the items on the list is `mvndst`. We plan to compile `_qmvnt.py`, and its `_qmvn` function should be able to eliminate the reliance on `_mvn.mvnun` (and resolve some open issues, like  gh-16142 and maybe gh-8367). The other use of `mvndst` is in [`gaussian_kde.integrate_box`](https://github.com/scipy/scipy/blob/4dff27ee9b129e6fea17f41a9ce11a6ad4f2162d/scipy/stats/_kde.py#L387C25-L387C29), where it calls `mvnun_weighted`. 

Would this functionality be difficult to incorporate into `qmvnt.py`? Would you be willing to add it?
